crates/passes/src/algorithms/inline_extern_func.rs

```diff
@@ -0,0 +1,63 @@
+use tohdl_ir::graph::CallNode;
+use tohdl_ir::graph::Node;
+use tohdl_ir::graph::NodeIndex;
+use tohdl_ir::graph::ReturnNode;
+use tohdl_ir::graph::CFG;
+
+/// Inlines an external function
+pub fn inline_extern_func<'a>(extern_node: NodeIndex, caller: &mut CFG, callee: &CFG) {
+    let old_to_new_idx = CFG::merge_graph(caller, callee);
+
+    let callee_exits = CFG::find_exits(callee).collect::<Vec<_>>();
+    let callee_exits = callee_exits
+        .into_iter()
+        .map(|x| old_to_new_idx.get(&x).unwrap())
+        .collect::<Vec<_>>();
+    let callee_entry = old_to_new_idx.get(&callee.entry).unwrap();
+
+    let extern_preds = caller.preds(extern_node).collect::<Vec<_>>();
+    let extern_succs = caller.succs(extern_node).collect::<Vec<_>>();
+    assert_eq!(
+        extern_succs.len(),
+        1,
+        "External node should only have one child being the func node"
+    );
+
+    // Connect call to entry
+    for pred in extern_preds {
+        let edge = caller.get_edge(pred, extern_node).unwrap();
+        caller.add_edge(pred, callee_entry.clone(), edge.clone());
+    }
+
+    // Connect exits to func node
+    for succ in extern_succs {
+        for exit in &callee_exits {
+            let edge = caller.get_edge(extern_node, succ).unwrap();
+            caller.add_edge((*exit).clone(), succ.clone(), edge.clone());
+        }
+    }
+
+    caller.rmv_node(extern_node);
+
+    // Replace exit return nodes with call nodes
+    for exit in callee_exits {
+        if let Some(node) = ReturnNode::concrete(caller.get_node(*exit)) {
+            let args = node
+                .values
+                .clone()
+                .into_iter()
+                .map(|x| match x {
+                    tohdl_ir::expr::Expr::Var(v) => v,
+                    _ => panic!("Expected VarExpr {}", x),
+                })
+                .collect();
+            let call_node = CallNode { args };
+            caller.replace_node(*exit, call_node)
+        } else {
+            panic!(
+                "Expected return node for function exit {}",
+                caller.get_node(*exit)
+            );
+        }
+    }
```

Comment:
Split into 3 functions?

Reply:
I don't see how splitting it into 3 functions would improve clarity. Those new functions would each have a `for` loop. Instead, I've moved some of the nested code into a separate function.